### PR TITLE
Refactor/programm membership enrichment

### PR DESCRIPTION
**Pull request by Judge40**
Refactor the programme membership enrichment, there should only be a
single entry point for enrichment - the ProgrammeMembership record.
All other related record types (Programme, Curriculum and
CurriculumMembership) are expanded in to the ProgrammeMembership queue
so each can be processed individually.
Using the power of the aggregated DTOs and mapping functions the
enrichment can be simplified significantly.

TIS21-4172
TIS21-2399

### CODE REVIEW

**Comment by Judge40** [comment]
@Health-Education-England/trainee-backend I've created this as a draft as there is still no handling for deleting PMs from the trainee profile. Currently there is no endpoint in `tis-trainee-details` to delete a single PM, so that needs adding first.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/facade/ProgrammeMembershipEnricherFacade.java [hunk 1]
```diff
@@ -187,558 +71,96 @@ public void enrich(Programme programme) {
...
-        }
-      }
-    }
-
-    if (doCurriculumEnrich) {
-      String curriculumId = getCurriculumId(programmeMembership);
-
-      if (curriculumId != null) {
+    if (curriculumMemberships.isEmpty()) {
+      curriculumMembershipService.requestForProgrammeMembership(programmeMembershipUuid);
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/facade/ProgrammeMembershipEnricherFacade.java:105]: I realise it makes no effective difference in what is returned, but for clarity might it be worthwhile to include a `dataRequested = true` statement here?

File: src/main/java/uk/nhs/hee/tis/trainee/sync/mapper/AggregateMapper.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+   *
+   * @param programmeMembership The programme membership to aggregate from.
+   * @param programme           The programme to aggregate from.
+   * @param curricula           The curricula to aggregate from.
+   * @return The aggregated programme membership DTO.
+   */
+  @Mapping(target = "tisId", source = "programmeMembership.uuid")
+  @Mapping(target = "personId", source = "programmeMembership.personId")
+  @Mapping(target = "programmeMembershipType",
+      source = "programmeMembership.programmeMembershipType")
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/mapper/AggregateMapper.java:80]: Is it obvious why some of these mappings use *.data.* and others don't? I'm not quite clear on the rationale, e.g. `curriculumMembership.data.curriculumStartDate` above, but `programmeMembership.programmeStartDate` below

Comment (Judge40) [reply src/main/java/uk/nhs/hee/tis/trainee/sync/mapper/AggregateMapper.java:80]: `ProgrammeMembership` was converted to a proper POJO in #351, so doesn't have the data map anymore.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/event/CurriculumMembershipEventListenerTest.java [hunk 3]
```diff
@@ -104,41 +187,97 @@ void shouldNotFindAndCacheProgrammeMembershipIfInCacheBeforeDelete() {
...
+  @Test
+  void shouldNotRequestMissingProgrammeMembershipAfterDelete() {
+    CurriculumMembership curriculumMembership = new CurriculumMembership();
+    String uuid = UUID.randomUUID().toString();
+    curriculumMembership.getData().put("programmeMembershipUuid", uuid);
+
+    when(cache.get("1", CurriculumMembership.class)).thenReturn(curriculumMembership);
+
+    ProgrammeMembership programmeMembership = new ProgrammeMembership();
+    programmeMembership.setUuid(UUID.fromString(uuid));
```
Comment (ReubenRobertsHEE) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/event/CurriculumMembershipEventListenerTest.java:270]: We don't need the two lines above I think

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/CurriculumMembershipSyncServiceTest.java [hunk 4]
```diff
@@ -317,4 +353,80 @@ void shouldThrowAnExceptionIfNotJsonProcessingException() throws JsonProcessingE
...
+    service.requestForProgrammeMembership(ID);
+    verify(dataRequestService, times(2))
+        .sendRequest("CurriculumMembership", whereMapPmUuid1);
+  }
+
+  @Test
+  void shouldSendRequestForProgrammeMembershipWhenRequestedDifferentIds()
+      throws JsonProcessingException {
+    service.requestForProgrammeMembership(ID);
+    service.requestForProgrammeMembership("ID2");
```
Comment (ReubenRobertsHEE) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/service/CurriculumMembershipSyncServiceTest.java:397]: Should this not be ID_2 ?

**Comment by ReubenRobertsHEE** [comment]
Apart from the nit-picks above, this looks great - really cleans-up the handling of programme memberships! 🌮

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/CurriculumMembershipSyncServiceTest.java (hunk 4 above)
Comment (Judge40) [reply src/test/java/uk/nhs/hee/tis/trainee/sync/service/CurriculumMembershipSyncServiceTest.java:397]: 🙈 you caught my bad copy paste

**Comment by Judge40** [comment]
> Apart from the nit-picks above, this looks great - really cleans-up the handling of programme memberships! 🌮

❤️ thanks for reviewing, I'll address the comments alongside the missing delete functionality

File: src/main/java/uk/nhs/hee/tis/trainee/sync/facade/ProgrammeMembershipEnricherFacade.java (hunk 1 above)
Comment (Judge40) [reply src/main/java/uk/nhs/hee/tis/trainee/sync/facade/ProgrammeMembershipEnricherFacade.java:105]: I didn't like the idea of adding a line of code that didn't really do anything, but I figured I could attach some logging to it and get both clarity and value.

```
if (curriculumMemberships.isEmpty()) {
  curriculumMembershipService.requestForProgrammeMembership(programmeMembershipUuid);
  dataRequested = true;
} else {
  ...
}

if (dataRequested) {
  log.info("blah blah skipped due to missing data.");
}

return dataRequested ? List.of() : aggregatedCurriculumMemberships;
```

File: src/test/java/uk/nhs/hee/tis/trainee/sync/event/ProgrammeMembershipEventListenerTest.java [hunk 5]
```diff
@@ -54,4 +88,73 @@ void shouldCallEnricherAfterSave() {
...
+    document.append("_id", ID);
+    AfterDeleteEvent<ProgrammeMembership> event = new AfterDeleteEvent<>(document, null, null);
+
+    listener.onAfterDelete(event);
+
+    verifyNoInteractions(tcsService);
+  }
+
+  @Test
+  void shouldSyncDeleteWhenProgrammeMembershipNotInCacheAfterDelete() {
```
Comment (ReubenRobertsHEE) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/event/ProgrammeMembershipEventListenerTest.java:137]: copy-pasta? shouldSyncDeleteWhenProgrammeMembership **Not** InCacheAfterDelete() 😁

File: src/test/java/uk/nhs/hee/tis/trainee/sync/event/ProgrammeMembershipEventListenerTest.java [hunk 6]
```diff
@@ -54,4 +88,73 @@ void shouldCallEnricherAfterSave() {
...
+    ArgumentCaptor<Record> recordCaptor = ArgumentCaptor.forClass(Record.class);
+    verify(tcsService).syncRecord(recordCaptor.capture());
+
+    Record programmeMembershipRecord = recordCaptor.getValue();
+    assertThat("Unexpected operation.", programmeMembershipRecord.getOperation(), is(DELETE));
+    assertThat("Unexpected schema.", programmeMembershipRecord.getSchema(), is("tcs"));
+    assertThat("Unexpected table.", programmeMembershipRecord.getTable(),
+        is("ProgrammeMembership"));
+    assertThat("Unexpected TIS ID.", programmeMembershipRecord.getTisId(), is(ID.toString()));
+    assertThat("Unexpected TIS ID.", programmeMembershipRecord.getData().get("personId"), is("40"));
```
Comment (ReubenRobertsHEE) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/event/ProgrammeMembershipEventListenerTest.java:158]: Nit-pick: maybe "Unexpected person ID"

**Comment by ReubenRobertsHEE** [comment]
Looks solid to me 🥇  It will be nice to test the end-to-end PM delete with trainee-details as well.

**Review by DorisWongNHS — APPROVED** [review]

**Pull request merged**